listing open games on home page

shaunshapiro opened the pull request:
Simple display of available games. I suppose next step would be to allow user to click on any open game and be routed to join that game..

bluem00se committed:
listing open games on home page

shaunshapiro commented:
@johnnyiller does this look ok?

johnnyiller commented on app/controllers/static_pages_controller.rb:
```diff
@@ -1,4 +1,5 @@
 class StaticPagesController < ApplicationController
 	def index
+		@games = Game.all
```
you should write a scope on the Game model so that it only returns games with white_uid nil or black_id nil

i'll give you a hint..

in the app/models/game.rb

implement the following method to return the proper scope.

def self.joinable
  # your code here
end

then in the controller you will call Game.joinable

then remove the if condition from the view...

shaunshapiro commented:
@johnnyiller should I be calling Game.joinable in the games controller or static pages controller? Because I'm still trying to list the games on the index.html.erb page, it seems to do something if I call Game.joinable in the static_pages_controller, though I'm still having a hard time figuring it out

shaunshapiro commented:
In app/models/game.rb I have: 

```
 def self.joinable  
      where(white_uid: nil).where(black_uid: nil).where_values.reduce(:or)
 end
```

In app/controllers/static_pages_controller:

```
 class StaticPagesController < ApplicationController
     def index
         @games = Game.joinable
     end
 end
```

And in index.html.erb:

```
 <% @games.each do |game| %>
     <%= game %>
     <br/>
 <% end %> 
```

This seems to display the games I have created in my database that are missing either a black or white uid in the form of arel tables. I can't figure out how to return just the game_name using this format though. I wrote the scope the way I did because I was having trouble figuring out how to chain an OR into the scope (black or white uid is nil) and this method seemed to get me closest to what I want so far.

shaunshapiro commented:
Just solved my problem. I realized that if I changed the scope to:

```
 where(white_uid: nil)
```

and returned game.game_name in index .html.erb it worked for returning nil white_uid games, so my issue was with chaining an OR into the scope. Wound up using the squeel gem to solve it pretty easily. Should have the code updated here momentarily.

bluem00se committed:
updated list_seats available to use a scoping method to return open games. Installed squeel gem.

bluem00se committed:
deleted a couple blank lines in app/controllers/games_controller.rb

johnnyiller commented:
squeel gem is not needed to do or conditions, you could leave it, but it seems like overkill to do something as simple as the following.

def self.joinable
where(Game.arel_table[:white_uid].eq(nil).or(Game.arel_table[:black_uid].eq(nil)))
end

Remember active record is backed by arel.  So if you have to do something that isn't supported out of the box you can look at https://github.com/rails/arel  for a bit of help.

johnnyiller commented:
That said, assuming this works you can merge it to production if you wish.

bluem00se committed:
Merge branch 'master' of github.com:Firehose-Team-Awesome/chess-game into list_seats_available

bluem00se committed:
edit .joinable function in game.rb

shaunshapiro commented:
I see, I tried something really similar to that but it didn't work. This line works perfectly. I removed squeel and updated the function. I'll merge it now as well.

jeffp2007 commented:
I'm a little confused. I just worked on listing open games and the joining feature. How is this feature different? Was I supposed to integrate the 'support joining a game' with this?
